torch/ao/ns/fx/n_shadows_utils.py

```diff
@@ -0,0 +1,893 @@
+        # subtract one because the first candidate is the floating point
+        # version of the subgraph
+        node_name_to_qconfig = \
+            list_of_node_name_to_qconfig[subgraph_candidate_idx - 1]
+        qconfig = node_name_to_qconfig[first_node.name]
+
+        # if no quantization is requested, skip
+        # TODO(future PR): deduplicate equivalent qconfigs that come from
+        #   different qconfig mapping objects
+        if qconfig is None:
+            return
+
+        qconfig_mapping = QConfigMapping().set_global(qconfig)
+
+        # create a copy of the submodule, wrapped in a separate module
+        orig_mod_copy_wrapped = create_submodule_from_subgraph(
+            mt, first_node, last_node)
+
+        # add a logger to the end of this submodule
+        # get first and last nodes of the submodule
+        _add_logger_to_subgraph_wrapper(
+            orig_mod_copy_wrapped, subgraph_idx, subgraph_candidate_idx,
+            str(qconfig), OutputComparisonLogger, last_node, fqn)
+
+        # We need to set the loggers as non traceable to have them survive
+        # prepare_fx and convert_fx calls.
+        prepare_custom_config = PrepareCustomConfig()\
+            .set_non_traceable_module_classes([OutputLogger, OutputComparisonLogger])
+
+        # add a call to prepare_fx on the wrapper module
+        orig_mod_copy_wrapped = torch.ao.quantization.quantize_fx.prepare_fx(
+            orig_mod_copy_wrapped, qconfig_mapping, example_inputs=example_inputs,
+            prepare_custom_config=prepare_custom_config)
+
+        # attach the wrapper to the model
+        attr_name = _get_attr_wrapper_name(subgraph_idx, subgraph_candidate_idx)
+        assert not hasattr(mt, attr_name)
+        setattr(mt, attr_name, orig_mod_copy_wrapped)
+
+        # add a call to the wrapper module from the parent graph
+        with mt.graph.inserting_after(last_node):
+            # TODO(future PR): handle fusion patterns where non-first nodes
+            # need inputs
+
+            # pass in all node args and kwargs
+
+            # the first argument is always the reference output of the last
+            # node of this subgraph
+            new_args = [last_node]
+
+            for arg in first_node.args:
+                if isinstance(arg, Node):
+                    new_args.append(arg)
+                elif isinstance(arg, (list, tuple)) and len(arg) and isinstance(arg[0], Node):
+                    for inner_arg in arg:
+                        if isinstance(inner_arg, Node):
+                            new_args.append(inner_arg)
+
+            new_kwargs = {}
+            for name, old_kwarg in first_node.kwargs.items():
+                if isinstance(old_kwarg, Node):
+                    new_kwargs[name] = old_kwarg
+                elif isinstance(old_kwarg, (list, tuple)) and len(old_kwarg):
+                    for inner_old_kwarg in old_kwarg:
+                        new_args.append(inner_old_kwarg)
```

Comment:
also not sure what this does, can we just pass around `first_node.kwargs` directly?

Reply:
happy to address in a future PR